lib/ansible/modules/remote_management/manageiq/manageiq_provider.py

```diff
@@ -0,0 +1,523 @@
+
+    def create_provider(self, name, provider_type, endpoints, zone_id, region):
+        """ Creates the user in manageiq.
+
+        Returns:
+            the created user id, name, created_on timestamp,
+            updated_on timestamp, userid and current_group_id.
+        """
+        # clean nulls, we do not send nulls to the api
+        endpoints = delete_nulls(endpoints)
+
+        # try to create a new provider
+        try:
+            url = '%s/providers' % (self.api_url)
+            result = self.client.post(
+                url,
+                name=name,
+                type=supported_providers()[provider_type]['class_name'],
+                zone={'id': zone_id},
+                provider_region=region,
+                connection_configurations=endpoints,
+            )
+        except Exception as e:
+            self.module.fail_json(msg="failed to create provider %s: %s" % (name, str(e)))
+
+        return dict(
+            changed=True,
+            msg="successfully created the provider %s: %s" % (name, result['results']))
+
+
+def main():
+    zone_id = None
+    endpoints = []
+
+    module = AnsibleModule(
+        argument_spec=dict(
+            manageiq_connection=dict(required=True, type='dict',
+                                     options=manageiq_argument_spec()),
+            state=dict(choices=['absent', 'present'], default='present'),
+            name=dict(required=True),
+            zone=dict(default='default'),
+            region=dict(),
+            type=dict(choices=supported_providers().keys()),
+            endpoints=dict(type='dict', options=endpoint_list_spec()),
+        ),
+        required_if=[
+            ('state', 'present', ['endpoints'])],
+    )
+
+    name = module.params['name']
+    zone_name = module.params['zone']
+    provider_type = module.params['type']
+    raw_endpoints = module.params['endpoints']
+    region = module.params['region']
+    state = module.params['state']
+
+    manageiq = ManageIQ(module)
+    manageiq_provider = ManageIQProvider(manageiq)
+
+    provider = manageiq_provider.provider(name)
+
+    # provider should not exist
+    if state == "absent":
+        # if we have a provider, delete it
+        if provider:
+            res_args = manageiq_provider.delete_provider(provider)
+        # if we do not have a provider, nothing to do
+        else:
+            res_args = dict(
+                changed=False,
+                msg="provider %s: does not exist in manageiq" % (name))
+
+    # provider shoult exist
+    if state == "present":
+        # get data user did not explicitly give
+        if zone_name:
+            zone_id = manageiq_provider.zone_id(zone_name)
+
+        # if we do not have a provider_type, use the current provider_type
+        if provider and not provider_type:
+            provider_type = manageiq_provider.class_name_to_type(provider['type'])
```

Comment:
p.s. added checks for null or bad provider type, checked ... this can happen if we manually tweak the type in the DB, and then send an edit request without explicit type :-)